prereise/gather/hiflddata/data_trans.py

```diff
@@ -0,0 +1,583 @@
+            line["geometry"]["coordinates"][0][-1][1],
+            line["geometry"]["coordinates"][0][-1][0],
+        )
+        candidate_coordinates = sub_name_dict.get(
+            line["properties"]["SUB_1"]
+        ) + sub_name_dict.get(line["properties"]["SUB_2"])
+
+        sub1 = min(
+            candidate_coordinates, key=lambda p: compute_geo_dist(p, start_coord)
+        )
+        sub2 = min(candidate_coordinates, key=lambda p: compute_geo_dist(p, end_coord))
+
+        dist = compute_geo_dist(start_coord, end_coord)
+
+        lines.append(
+            [
+                line_id,  # line id
+                line_type,  # line type
+                sub_by_coord_dict.get(sub1)[0],  # from substation id
+                sub_by_coord_dict.get(sub1)[1],  # from substation name
+                sub_by_coord_dict.get(sub2)[0],  # to substation id
+                sub_by_coord_dict.get(sub2)[1],  # to substation name
+                dist,  # distance between start point and end point of the line.
+            ]
+        )
+
+        if sub1 == sub2:
+            continue
+        n_dict[sub1].append(sub2)
+        n_dict[sub2].append(sub1)
+
+    df_lines = pd.DataFrame(
+        lines,
+        columns=[
+            "branch_id",
+            "line_type",
+            "from_bus_id",
+            "from_bus_name",
+            "to_bus_id",
+            "to_bus_name",
+            "length_in_mile",
+        ],
+    )
+    return df_lines, n_dict
+
+
+def line_from_csv(t_csv):
+    """Create dict to store all the raw transmission line csv data
+
+    :param str t_csv: path of the HIFLD transmission csv file
+    :return: (*dict*) -- a dict mapping the transmission ID to its raw parameters.
+    """
+
+    raw_data = pd.read_csv(t_csv)
+    raw_data["ID"] = raw_data["ID"].astype("str")
+    raw_lines = raw_data.set_index("ID").to_dict()
+    return raw_lines
+
+
+def meter_to_mile(dist):
+    """Calculate the mile given the distance in meter
+
+    :param float dist: length of the distance between two substations
+    :return: (*float*) -- distance in mile unit.
+    """
+
+    return dist / 1609.34
+
+
+def compute_geo_dist(sub1, sub2):
+    """Compute the haversine geo distance of two substations
+
+    :param tuple sub1: (latitude, longitude) of substation1
+    :param tuple sub2: (latitude, longitude) of substation2
+    :return: (*float*) -- distance between two geo coordinates.
+    """
+
+    return haversine(sub1, sub2, Unit.MILES)
+
+
+def graph_of_net(n_dict):
+    """Create the graph for the HIFLD transmission network
+
+    :param dict n_dict: dict of substation's neighbors as the edge inside the Graph
+    :return: (*networkx.Graph*) -- Graph generated by networkx.
+    """
+    nodes = list(n_dict.keys())
+    graph = nx.Graph()
+    any(graph.add_node(node) for node in nodes)
+    any(
+        graph.add_edge(node, neighbor)
+        for node, neighbors in n_dict.items()
+        for neighbor in neighbors
+    )
```

Comment:
Why do we evaluate `any` on these repeated graph method calls?

Reply:
Using `any` here is quite confusing and it was some initial attempt to refactor some lines into more Python fashion. I think it is better to revert this back using for loop. Will also switch to graph.add_nodes_from to bulk add all the nodes.